cmd/gaia/app/export.go

```diff
@@ -0,0 +1,156 @@
+package app
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth"
+	distr "github.com/cosmos/cosmos-sdk/x/distribution"
+	"github.com/cosmos/cosmos-sdk/x/gov"
+	"github.com/cosmos/cosmos-sdk/x/mint"
+	"github.com/cosmos/cosmos-sdk/x/slashing"
+	stake "github.com/cosmos/cosmos-sdk/x/stake"
+	abci "github.com/tendermint/tendermint/abci/types"
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+// export the state of gaia for a genesis file
+func (app *GaiaApp) ExportAppStateAndValidators(forZeroHeight bool) (
+	appState json.RawMessage, validators []tmtypes.GenesisValidator, err error) {
+
+	// as if they could withdraw from the start of the next block
+	ctx := app.NewContext(true, abci.Header{Height: app.LastBlockHeight()})
+
+	if forZeroHeight {
+		prepForZeroHeightGenesis(ctx)
+	}
+
+	// iterate to get the accounts
+	accounts := []GenesisAccount{}
+	appendAccount := func(acc auth.Account) (stop bool) {
+		account := NewGenesisAccountI(acc)
+		accounts = append(accounts, account)
+		return false
+	}
+	app.accountKeeper.IterateAccounts(ctx, appendAccount)
+
+	genState := NewGenesisState(
+		accounts,
+		auth.ExportGenesis(ctx, app.feeCollectionKeeper),
+		stake.ExportGenesis(ctx, app.stakeKeeper),
+		mint.ExportGenesis(ctx, app.mintKeeper),
+		distr.ExportGenesis(ctx, app.distrKeeper),
+		gov.ExportGenesis(ctx, app.govKeeper),
+		slashing.ExportGenesis(ctx, app.slashingKeeper),
+	)
+	appState, err = codec.MarshalJSONIndent(app.cdc, genState)
+	if err != nil {
+		return nil, nil, err
+	}
+	validators = stake.WriteValidators(ctx, app.stakeKeeper)
+	return appState, validators, nil
+}
+
+// prepare for fresh start at zero height
+func prepForZeroHeightGenesis(ctx sdk.Context) {
```

Comment:
(I just rearranged some code here to reduce indentation/increase clarity.)
Okay probably my biggest comment for this PR is most of the stuff that's in this function should actually be moved to export-genesis functionality of their respective modules - most of the code here is module specific, not gaia specific. I shouldn't have to fully grasp all the elements of each module if I was constructing my own blockchain as would be required if we leave this as is. 

This function should ideally be reduced to a few lines of code simply calling "prepForZeroHeightGenesis" functionality for each module. Then we could actually eliminate this function altogether and move it back up to the above functionality as it would no longer be a massive indentation

Reply:
This would apply to both the invariance checks as well as prep functionality

Reply:
Hmm. I agree in principle. I do wonder whether zero-height export is something we want to pursue as a long-term feature - initially my thought was to do this just for gaia-9002 / GoS as a quick hack, and then implement "start-Tendermint-at-height-n" as the real long-term solution (at which point we could delete this code, since we would no longer need the feature to export for height zero).

Reply:
yeah I mean if we're going to throw this away then of course no point wasting energy. And reflecting, I think we probably should. K nm - let's leave